File: longest_word._dfs.py
# Time Complexity : 
#      For Insertion: O(k*N) where k is averange length of the given words and N is number of words given
#      For Finding Longest Word: O(k) where k is length of the longest word
# Space Complexity : O(k * N) where k is averange length of the word and N is number of words given
# Did this code successfully run on Leetcode : Yes
# Any problem you faced while coding this : No

class TrieNode():
    def __init__(self):
        self.children = [None for _ in range(26)]
        self.isEnd = False
        self.character = ""

class Solution:
    def __init__(self):
        self.root = TrieNode()
        self.result = []
    def longestWord(self, words: List[str]) -> str:
        for word in words:
            self.insertWord(word)
        
        curr = self.root
        self.dfs(curr, [])
        
        return "".join(self.result)
    
    def dfs(self, curr, currLongest):
        #base
        if len(self.result) < len(currLongest):
            self.result = currLongest[:]
        #logic
        for i in range(26):
            if curr.children[i] is not None and curr.children[i].isEnd == True:
                currLongest.append(curr.children[i].character)
                
                self.dfs(curr.children[i], currLongest)
                
                currLongest.pop()
                
            
        
        
    def insertWord(self, word):
        curr = self.root
        for l in word:
            if curr.children[ord(l) - ord('a')] == None:
                curr.children[ord(l) - ord('a')] = TrieNode()
            curr = curr.children[ord(l) - ord('a')]
            curr.character = l
            
        curr.isEnd = True
        

File: LongestWordDFS.java
public class LongestWordDFS {

        //DFS - Trie
        public static class TrieNode {

            TrieNode[] children;
            boolean isEnd;

            public TrieNode() {

                this.children = new TrieNode[26];
            }

        }

        public void insertInTrie(TrieNode root, String word) {

            TrieNode curr = root;

            for(int i = 0; i < word.length(); i++) {

                char c = word.charAt(i);

                if(curr.children[c - 'a'] == null) {

                    curr.children[c - 'a'] = new TrieNode();
                }

                curr = curr.children[c - 'a'];
            }

            curr.isEnd = true;
        }

        String longest;

        public String longestWord(String[] words) {

            // initiate empty longest string
            this.longest = "";

            // make trie of given words
            TrieNode rootMain = new TrieNode();

            for(String word: words) { //O(n*l)

                insertInTrie(rootMain, word);
            }

            // run dfs
            backtrackDFS(rootMain, new StringBuilder()); // O(n*l)

            // output
            return longest;
        }

        public void backtrackDFS(TrieNode currDfs, StringBuilder sb) {

            //base
            //update the longest if current StringBuilder is longer
            if(sb.length() > longest.length()) {

                //longest = new StringBuilder(sb).toString();
                longest = new String(sb);
            }

            //logic
            //dfs on all 26 array Trie Nodes of currDfs Trie Node
            for(int i = 0; i < 26; i++) {

                // if a word exists, then ARB
                if(currDfs.children[i] != null && currDfs.children[i].isEnd) {

                    //action
                    sb.append((char)('a' + i));

                    //recurse
                    backtrackDFS(currDfs.children[i], sb);

                    System.out.println(sb);

                    //backtrack
                    sb.deleteCharAt(sb.length() - 1);
                }
            }
        }

}

/*
TIME COMPLEXITY = O(n*l)
n - number of words
l - length of each word

build tree and search - n*l time

SPACE COMPLEXITY = O(n*l)

trie size - n*l
*/